### PR TITLE
Modernize e2e - OKTA-203974

### PR DESCRIPTION
**Pull request by shuowu**
## Description:

This PR modernizes the E2E process in the widget repo. While the refactor is in progress, bacon will run both `e2e` and `e2e-wdio` script at the same time, and we will only remove the old tests (under test/e2e) when the migration is fully done.

Changes in this PR:
* add a new react test harness app under test/app
* rewrite old protractor based tests in webdriveIO format 
* new tests is configured for chrome only on bacon

**Edit:**
PR Updated based on discussion on Monday:
* Removed react dep from test harness app, now it's written in pure typescript. 
* Replaced mocha with jasmin

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-203974](https://oktainc.atlassian.net/browse/OKTA-203974)

### CODE REVIEW

**Review by haishengwu-okta — COMMENTED** [review]
Why another framework instead of testcafe ...

**Comment by shuowu** [comment]
@haishengwu-okta Good question, WebdriveIO is added mostly to align with other E2E setups from other JS repos (okta-auth-js), and it provides better support for [multiple browser windows](https://testcafe.io/documentation/402841/guides/advanced-guides/multiple-browser-windows#supported-browsers) testing scenarios. Currently, testcafe tests only runs against mock responses, probably they can just be taken as two different tests.

But, I don't have strong opinion on which framework to pick, open to suggestions.

File: package.json [hunk 1]
```diff
@@ -59,30 +60,41 @@
...
   },
   "devDependencies": {
-    "@babel/core": "^7.10.3",
+    "@babel/core": "^7.15.5",
     "@babel/plugin-transform-modules-commonjs": "^7.10.1",
     "@babel/plugin-transform-runtime": "^7.10.3",
-    "@babel/preset-env": "^7.10.3",
+    "@babel/preset-env": "^7.15.6",
+    "@babel/preset-react": "^7.14.5",
+    "@babel/register": "^7.15.3",
```
Comment (jmelberg-okta) [inline package.json:72]: Are both of these dependencies required to run wdio? Seems a bit odd if so.

File: package.json [hunk 2]
```diff
@@ -59,30 +60,41 @@
...
     "@babel/runtime-corejs3": "^7.10.3",
     "@okta/eslint-plugin-okta": "0.4.0",
     "@okta/okta-auth-js": "^5.5.0",
     "@okta/okta-idx-js": "0.20.0",
     "@sindresorhus/to-milliseconds": "^1.0.0",
+    "@wdio/cli": "^7.14.1",
+    "@wdio/junit-reporter": "^7.15.0",
+    "@wdio/local-runner": "^7.14.1",
+    "@wdio/mocha-framework": "^7.14.1",
```
Comment (jmelberg-okta) [inline package.json:81]: IIRC we have plans to migrate off of mocha to another framework. Do you have a ticket I could follow to track that work?

File: test/app/src/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
```
Comment (jmelberg-okta) [inline test/app/src/index.js:1]: So we now require React to be installed to run tests locally? This seems like a huge degrade in developer experience - as most devs do not add/edit e2e tests.

Can you explain the rational for using React versus our prior pure JS implementation?

File: test/e2e-wdio/babel.config.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  'presets': [
+    [
+      '@babel/preset-env',
+      {
+        'targets': {
+          'node': '14'
```
Comment (jmelberg-okta) [inline test/e2e-wdio/babel.config.js:7]: Please make sure this maps to our `engines` and what runs in CI.

File: test/e2e-wdio/page-objects/primary-auth.page.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { waitForLoad } from '../util/waitUtil';
+
+class PrimaryLoginPage {
```
Comment (pradeepdewda-okta) [inline test/e2e-wdio/page-objects/primary-auth.page.js:3]: we should name this `PrimaryAuthPage` as file name indicate

File: test/app/src/getOktaSignIn_CDN.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+declare global {
```
Comment (haishengwu-okta) [inline test/app/src/getOktaSignIn_CDN.ts:1]: Can we change the file naming convention like `OktaSignInCDN.ts` or any other you'd think of without mixing `_` and verbs.

**Review by haishengwu-okta — APPROVED** [review]
Question around naming then 🚀

File: test/e2e-wdio/README.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+**Note:** Before the repos is transformed into a yarn workspace, the test should be run in the root directory of the repo.
+
+// TODO: https://oktainc.atlassian.net/browse/OKTA-437720
+
+## Define environment variables
+
+See `test/e2e-wdio/env.defaults.js` for a list of all environment variables used by E2E tests. You can define variables in the shell (using `export VAR=value` in `~/.bash_profile` or similar) or place values in a `testenv` file at the root of this project. [dotenv docs](https://github.com/motdotla/dotenv#dotenv)
+
+For all E2E tests to pass locally, you will need to define these values. You will need a test org and a FB user.
```
Comment (pradeepdewda-okta) [inline test/e2e-wdio/README.md:11]: what is FB user here ?

File: test/e2e-wdio/README.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+The test org should have a configured SPA app with following login redirect callbacks:
```
Comment (pradeepdewda-okta) [inline test/e2e-wdio/README.md:13]: can we explain this a little bit more in the README.. like why ?

File: test/app/src/getOktaSignIn_CDN.ts (hunk 6 above)
Comment (pradeepdewda-okta) [reply test/app/src/getOktaSignIn_CDN.ts:1]: change to `getOktaSingInWIdgetFromCDN`

File: test/app/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# E2E test app
+
+This is a react test harness app for widget E2E testing. This app provides UI components for test runner to update widget config and render updated widget based on top of that.
```
Comment (pradeepdewda-okta) [inline test/app/README.md:3]: still say `react test harness app` ?

File: test/app/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
+const webpack = require('webpack');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const path = require('path');
+const ENV = require('../e2e/env');
+ENV.config();
+
+const DEV_SERVER_PORT = 3000;
```
[the remark this answers is not in the document]
Comment (shuowu) [reply test/app/webpack.config.js:7]: Let's keep using 3000, as it need to match the config from the test org which has been pre-configured.

File: .bacon.yml [hunk 11]
```diff
@@ -41,6 +41,13 @@ test_suites:
       script_name: e2e
       criteria: MERGE
       queue_name: small
+    - name: e2e-wdio
+      script_path: /root/okta/okta-signin-widget/scripts
+      sort_order: '5'
+      timeout: '10'
```
[the remark this answers is not in the document]
Comment (shuowu) [reply .bacon.yml:47]: it's good for now, how about increase the timeout when needed? give it relative large number may keep the devs waiting on the bacon task when issues happen.

**Pull request closed without merging**